File: 0040-combination-sum-ii/0040-combination-sum-ii.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
  const results = [];
  candidates.sort((a,b) => a-b);
  
  // Backtracking
  function searchPath(index, path, target) {
    if (target < 0) return;
    if (target === 0) return results.push(path.slice());
    
    for (let i = index; i < candidates.length; i++) {
      if (i > index && candidates[i] === candidates[i - 1]) continue;
      path.push(candidates[i]);
      searchPath(i + 1, path, target - candidates[i]); 
      path.pop();
    }
  };
  
  searchPath(0, [], target);
  return results;
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        results = []
        
        def searchPath(index, path, target):
            # base cases
            if target < 0: 
                return
            if target == 0:
                return results.append(path[:])
            
            for i in range(index, len(candidates)):
                if i > index and candidates[i] == candidates[i-1]: 
                    continue
                path.append(candidates[i])
                searchPath(i + 1, path, target - candidates[i])
                path.pop()
        
        searchPath(0, [], target)
        return results